### ISSUE TITLE
PyPI package names

### ISSUE THREAD

**Issue by lewisjared** [CMIP-REF/cmip-ref#22]
## The problem
We should start publishing to pypi to at bare minimum reserve our package naming, but to also ensure the publication workflow is 
fit for purpose.

I'd prefer to publish the following to PyPI:
* ref (Compute engine)
* ref-core (Shared functionality needed by metric providers)

Sadly, the `ref` pypi package is already claimed, but it has been unused for the last 6 years. I've sent an email to address associated with the project to see if they would be willing to assign the ownership of the ref PyPi to myself, but who knows if that will pan out.

We should make a backup plan. 

Perhaps we could publish as `cmip-ref` and `cmip-ref-core`. This results in a long name for the metrics packages (`cmip-ref-metrics-esmvaltool`). Not a major as these don't need to be published to pypi, but 🤷 

Another option would be to rename the `ref` package to something else (`ref-engine`?)

Any other ideas?

In #20 I've added a bump, release and deploy workflows for publishing new releases. 
This is untested for this repo, but we (@mikapfl  and I) have used this flow with success in other projects.

In addition, before publishing we need to resolve:
* [ ] #21 
* [x] #12 

## Definition of "done"
* [ ] Choose naming convention
* [ ] Rename packages if needed
* [ ] Bump a new package and deploy to pypi

**Comment by lewisjared** [comment]
@lee1043 @nocollier @bouweandela @climate-dude @hb326 Any other suggested package names?

**Comment by lee1043** [comment]
Hi @lewisjared, thanks for the discussion. I think I like `cmip-ref` most, as it is simple and short. Having more than one dash seems not very ideal to me. I also think `ref` without `cmip` might be too general.

**Comment by mikapfl** [comment]
Yeah, `ref` without anything else is probably too short for a global namespace like PyPIs. So `cmip-ref` sounds good. `cmip-ref-metrics-esmvaltool` is long, but also not too bad, I think. But I generally like long, overly descriptive names. (-;

**Comment by lewisjared** [comment]
Ok sounds like `cmip-ref` is the namespace we will use. Long package names aren't too bad as you only need to remember once when installing, the installed python packages drop the `cmip` prefix i.e. (`ref` and `ref_core`)

**Comment by bouweandela** [comment]
> ref (Compute engine)
  ref-core (Shared functionality needed by metric providers)

How about calling these `cmip_ref_compute` (or `cmip_ref_engine`) and `cmip_ref_metrics` so the names describe what's in the package? `core` is rather vague.

> the installed python packages drop the cmip prefix i.e. (`ref` and `ref_core`)

I would recommend against using a different name for the package and the imports, this is confusing to users. How about using a [namespace package](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/) to create some separation while maintaining a uniform look and feel for users? We could then do `import cmip_ref` or `import cmip_ref.core`. It may be desirable to use underscores in the package names to make them identical to the import statements, that will simplify things for users.

> `cmip-ref-metrics-esmvaltool` is long

We could potentially hide these long names from users by using extras, e.g. `pip install cmip_ref[esmvaltool, ilamb]` to install the metric packages.

**Comment by lewisjared** [comment]
> I install just cmip_ref, what is actually installed?

This would be the install target for end users to install the framework (by default without any providers). They don't need to know about the different packages, which is an implementation detail. Core (or whatever name we use) is only needed in metric provider packages. 

I haven't played with namespace packages before. Is there an example of a mature package using this pattern?

> I would recommend against using a different name for the package and the imports, this is confusing to users.

This is an interesting debate. PyPi [normalises to use dashes](https://packaging.python.org/en/latest/specifications/name-normalization/#name-normalization) so you could use `pip install cmip_ref` or `pip install cmip-ref`. Personally, I prefer to differentiate distribution packages from imports. It isn't a one to one link.

It's also not a hill I'm going to die on 😁 We would then also have to rename the repo

> We could potentially hide these long names from users by using extras, e.g. pip install cmip_ref[esmvaltool, ilamb] to install the metric packages.

Good point (especially for the "sanctioned" metric providers). That might be a somewhat niche use-case as that would install the provider into the current venv which may/may not be solvable. Another pattern would be to use the cli to install providers:`ref provider install esmvaltool` which could create a new conda env, register the provider/metrics with the database and set up any cross-process infra that is needed. 

I haven't thought through the install process, but it will be solvable.

**Comment by lewisjared** [comment]
Discussed with @bouweandela and @mikapfl. We will migrate to `cmip_ref` and `cmip_ref_core` and publish under the same name on pypi. The `cmip_ref` package might also provide a few extras to support different ways of configuring the REF.
